// ternoa-js/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Errors = exports.WaitUntil = exports.chainConstants = exports.chainQuery = exports.txEvent = exports.txActions = exports.txPallets = void 0;
var txPallets;
(function (txPallets) {
    txPallets["assets"] = "assets";
    txPallets["auction"] = "auction";
    txPallets["marketplace"] = "marketplace";
    txPallets["nft"] = "nft";
    txPallets["rent"] = "rent";
    txPallets["utility"] = "utility";
    txPallets["balances"] = "balances";
    txPallets["associatedAccounts"] = "associatedAccounts";
    txPallets["system"] = "system";
    txPallets["tee"] = "tee";
    txPallets["transmissionProtocols"] = "transmissionProtocols";
})(txPallets || (exports.txPallets = txPallets = {}));
var txActions;
(function (txActions) {
    txActions["create"] = "create";
    txActions["transfer"] = "transfer";
    txActions["transferAll"] = "transferAll";
    txActions["transferKeepAlive"] = "transferKeepAlive";
    txActions["batch"] = "batch";
    txActions["batchAll"] = "batchAll";
    txActions["forceBatch"] = "forceBatch";
    // nft
    txActions["addSecret"] = "addSecret";
    txActions["createNft"] = "createNft";
    txActions["createSecretNft"] = "createSecretNft";
    txActions["burnNft"] = "burnNft";
    txActions["transferNft"] = "transferNft";
    txActions["delegateNft"] = "delegateNft";
    txActions["setRoyalty"] = "setRoyalty";
    txActions["addNftToCollection"] = "addNftToCollection";
    txActions["createCollection"] = "createCollection";
    txActions["limitCollection"] = "limitCollection";
    txActions["closeCollection"] = "closeCollection";
    txActions["burnCollection"] = "burnCollection";
    txActions["setCollectionOffchaindata"] = "setCollectionOffchaindata";
    txActions["setNftMintFee"] = "setNftMintFee";
    // capsule
    txActions["convertToCapsule"] = "convertToCapsule";
    txActions["createCapsule"] = "createCapsule";
    txActions["revertCapsule"] = "revertCapsule";
    txActions["setCapsuleOffchaindata"] = "setCapsuleOffchaindata";
    txActions["notifyEnclaveKeyUpdate"] = "notifyEnclaveKeyUpdate";
    // rent
    txActions["createContract"] = "createContract";
    txActions["cancelContract"] = "cancelContract";
    txActions["acceptRentOffer"] = "acceptRentOffer";
    txActions["acceptSubscriptionTerms"] = "acceptSubscriptionTerms";
    txActions["changeSubscriptionTerms"] = "changeSubscriptionTerms";
    txActions["rent"] = "rent";
    txActions["makeRentOffer"] = "makeRentOffer";
    txActions["retractRentOffer"] = "retractRentOffer";
    txActions["revokeContract"] = "revokeContract";
    // transmission protocols
    txActions["addConsent"] = "addConsent";
    txActions["removeTransmissionProtocol"] = "removeTransmissionProtocol";
    txActions["resetTimer"] = "resetTimer";
    txActions["setTransmissionProtocol"] = "setTransmissionProtocol";
    // marketplace
    txActions["buyNft"] = "buyNft";
    txActions["createMarketplace"] = "createMarketplace";
    txActions["listNft"] = "listNft";
    txActions["unlistNft"] = "unlistNft";
    txActions["setMarketplaceConfiguration"] = "setMarketplaceConfiguration";
    txActions["setMarketplaceKind"] = "setMarketplaceKind";
    txActions["setMarketplaceOwner"] = "setMarketplaceOwner";
    txActions["setMarketplaceMintFee"] = "setMarketplaceMintFee";
    // auction
    txActions["createAuction"] = "createAuction";
    txActions["cancelAuction"] = "cancelAuction";
    txActions["endAuction"] = "endAuction";
    txActions["addBid"] = "addBid";
    txActions["removeBid"] = "removeBid";
    txActions["buyItNow"] = "buyItNow";
    txActions["claim"] = "claim";
    // tee
    txActions["submitMetricsServerReport"] = "submitMetricsServerReport";
    txActions["claimRewards"] = "claimRewards";
})(txActions || (exports.txActions = txActions = {}));
var txEvent;
(function (txEvent) {
    txEvent["ExtrinsicSuccess"] = "ExtrinsicSuccess";
    txEvent["ExtrinsicFailed"] = "ExtrinsicFailed";
    txEvent["BatchCompleted"] = "BatchCompleted";
    txEvent["BatchInterrupted"] = "BatchInterrupted";
    txEvent["nftsCreated"] = "Created";
    txEvent["nftsBurned"] = "Burned";
    txEvent["nftsTransfered"] = "Transfered";
    txEvent["MarketplaceCreated"] = "MarketplaceCreated";
})(txEvent || (exports.txEvent = txEvent = {}));
var chainQuery;
(function (chainQuery) {
    chainQuery["nftMintFee"] = "nftMintFee";
    chainQuery["secretNftMintFee"] = "secretNftMintFee";
    chainQuery["secretNftsOffchainData"] = "secretNftsOffchainData";
    chainQuery["nfts"] = "nfts";
    chainQuery["nextNFTId"] = "nextNFTId";
    chainQuery["nextCollectionId"] = "nextCollectionId";
    chainQuery["marketplaceMintFee"] = "marketplaceMintFee";
    chainQuery["account"] = "account";
    chainQuery["number"] = "number";
    chainQuery["collections"] = "collections";
    chainQuery["nextMarketplaceId"] = "nextMarketplaceId";
    chainQuery["marketplaces"] = "marketplaces";
    chainQuery["listedNfts"] = "listedNfts";
    // capsule
    chainQuery["capsuleMintFee"] = "capsuleMintFee";
    chainQuery["capsuleOffchainData"] = "capsuleOffchainData";
    // transmissionProtocols
    chainQuery["atBlockFee"] = "atBlockFee";
    chainQuery["atBlockWithResetFee"] = "atBlockWithResetFee";
    chainQuery["onConsentFee"] = "onConsentFee";
    chainQuery["onConsentAtBlockFee"] = "onConsentAtBlockFee";
    chainQuery["atBlockQueue"] = "atBlockQueue";
    chainQuery["transmissions"] = "transmissions";
    chainQuery["onConsentData"] = "onConsentData";
    // auction
    chainQuery["auctions"] = "auctions";
    chainQuery["deadlines"] = "deadlines";
    chainQuery["claims"] = "claims";
    // rent
    chainQuery["contracts"] = "contracts";
    chainQuery["queues"] = "queues";
    chainQuery["offers"] = "offers";
    // tee
    chainQuery["clusterData"] = "clusterData";
    chainQuery["enclaveData"] = "enclaveData";
    chainQuery["nextClusterId"] = "nextClusterId";
})(chainQuery || (exports.chainQuery = chainQuery = {}));
var chainConstants;
(function (chainConstants) {
    chainConstants["initialMintFee"] = "initialMintFee";
    chainConstants["initialSecretMintFee"] = "initialSecretMintFee";
    chainConstants["collectionSizeLimit"] = "collectionSizeLimit";
    chainConstants["existentialDeposit"] = "existentialDeposit";
    chainConstants["nftOffchainDataLimit"] = "nftOffchainDataLimit";
    chainConstants["collectionOffchainDataLimit"] = "collectionOffchainDataLimit";
    chainConstants["offchainDataLimit"] = "offchainDataLimit";
    chainConstants["accountSizeLimit"] = "accountSizeLimit";
    // auction
    chainConstants["auctionEndingPeriod"] = "auctionEndingPeriod";
    chainConstants["auctionGracePeriod"] = "auctionGracePeriod";
    chainConstants["bidderListLengthLimit"] = "bidderListLengthLimit";
    chainConstants["maxAuctionDelay"] = "maxAuctionDelay";
    chainConstants["minAuctionDuration"] = "minAuctionDuration";
    chainConstants["maxAuctionDuration"] = "maxAuctionDuration";
    chainConstants["parallelAuctionLimit"] = "parallelAuctionLimit";
    // rent
    chainConstants["actionsInBlockLimit"] = "actionsInBlockLimit";
    chainConstants["maximumContractAvailabilityLimit"] = "maximumContractAvailabilityLimit";
    chainConstants["maximumContractDurationLimit"] = "maximumContractDurationLimit";
    chainConstants["simultaneousContractLimit"] = "simultaneousContractLimit";
    // transmissionProtocols
    chainConstants["simultaneousTransmissionLimit"] = "simultaneousTransmissionLimit";
    chainConstants["maxConsentListSize"] = "maxConsentListSize";
    chainConstants["maxBlockDuration"] = "maxBlockDuration";
})(chainConstants || (exports.chainConstants = chainConstants = {}));
var WaitUntil;
(function (WaitUntil) {
    WaitUntil[WaitUntil["BlockInclusion"] = 0] = "BlockInclusion";
    WaitUntil[WaitUntil["BlockFinalization"] = 1] = "BlockFinalization";
})(WaitUntil || (exports.WaitUntil = WaitUntil = {}));
var Errors;
(function (Errors) {
    Errors["EXTRINSIC_FAILED"] = "EXTRINSIC_FAILED";
    Errors["EVENT_NOT_FOUND"] = "EVENT_NOT_FOUND";
    Errors["SEED_NOT_FOUND"] = "SEED_NOT_FOUND";
    Errors["PUBLIC_SEED_ADDRESS_NOT_FOUND"] = "PUBLIC_SEED_ADDRESS_NOT_FOUND";
    Errors["VALUE_MUST_BE_GREATER_THAN_0"] = "VALUE_MUST_BE_GREATER_THAN_0";
    Errors["VALUE_MUST_BE_DEFINED"] = "VALUE_MUST_BE_DEFINED";
    Errors["INSUFFICIENT_FUNDS"] = "INSUFFICIENT_FUNDS";
    Errors["API_NOT_INITIALIZED"] = "API_NOT_INITIALIZED";
    Errors["API_NOT_CONNECTED"] = "API_NOT_CONNECTED";
    Errors["TRANSACTION_NOT_IN_BLOCK"] = "TRANSACTION_NOT_IN_BLOCK";
    Errors["EXTRINSIC_NOT_FOUND"] = "EXTRINSIC_NOT_FOUND";
    Errors["OFFCHAIN_LENGTH_TOO_HIGH"] = "OFFCHAIN_LENGTH_TOO_HIGH";
    Errors["LIMIT_TOO_LOW"] = "LIMIT_TOO_LOW";
    Errors["LIMIT_TOO_HIGH"] = "LIMIT_TOO_HIGH";
    Errors["NFT_NOT_FOUND"] = "NFT_NOT_FOUND";
    Errors["COLLECTION_NOT_FOUND"] = "COLLECTION_NOT_FOUND";
    Errors["MUST_BE_A_NUMBER"] = "MUST_BE_A_NUMBER";
    Errors["URL_UNDEFINED"] = "URL_UNDEFINED";
    Errors["MUST_BE_PERCENTAGE"] = "MUST_BE_PERCENTAGE";
    Errors["NFT_CONVERSION_ERROR"] = "NFT_CONVERSION_ERROR";
    Errors["COLLECTION_CONVERSION_ERROR"] = "COLLECTION_CONVERSION_ERROR";
    Errors["MARKETPLACE_CONVERSION_ERROR"] = "MARKETPLACE_CONVERSION_ERROR";
    Errors["LISTED_NFT_CONVERSION_ERROR"] = "LISTED_NFT_CONVERSION_ERROR";
    Errors["IPFS_FILE_UPLOAD_ERROR"] = "IPFS_FILE_UPLOAD_ERROR";
    Errors["IPFS_METADATA_VALIDATION_ERROR"] = "IPFS_METADATA_VALIDATION_ERROR";
    Errors["BLOCK_NOT_FOUND_ON_CHAIN"] = "BLOCK_NOT_FOUND_ON_CHAIN";
    Errors["AUCTION_NFT_CONVERSION_ERROR"] = "AUCTION_NFT_CONVERSION_ERROR";
    Errors["RENT_NFT_CONVERSION_ERROR"] = "RENT_NFT_CONVERSION_ERROR";
    Errors["CLUSTER_CONVERSION_ERROR"] = "CLUSTER_CONVERSION_ERROR";
    Errors["ENCLAVE_CONVERSION_ERROR"] = "ENCLAVE_CONVERSION_ERROR";
    Errors["TRANSMISSION_PROTOCOL_CONVERSION_ERROR"] = "TRANSMISSION_PROTOCOL_CONVERSION_ERROR";
    Errors["TEE_CLUSTER_NOT_FOUND"] = "TEE_CLUSTER_NOT_FOUND";
    Errors["NEXT_TEE_CLUSTER_UNDEFINED"] = "NEXT_TEE_CLUSTER_UNDEFINED";
    Errors["TEE_CLUSTER_IS_EMPTY"] = "TEE_CLUSTER_IS_EMPTY";
    Errors["TEE_ENCLAVE_NOT_FOUND"] = "TEE_ENCLAVE_NOT_FOUND";
    Errors["TEE_ENCLAVE_NOT_AVAILBLE"] = "TEE_ENCLAVE_NOT_AVAILBLE";
    Errors["TEE_UPLOAD_ERROR"] = "TEE_UPLOAD_ERROR";
    Errors["TEE_RETRIEVE_ERROR"] = "TEE_RETRIEVE_ERROR";
    Errors["TEE_REMOVE_ERROR"] = "TEE_REMOVE_ERROR";
    Errors["TEE_ERROR"] = "TEE_ERROR";
    Errors["NOT_CORRECT_AMOUNT_TEE_PAYLOADS"] = "NOT_CORRECT_AMOUNT_TEE_PAYLOADS";
    Errors["NOT_CORRECT_AMOUNT_TEE_ENCLAVES"] = "NOT_CORRECT_AMOUNT_TEE_ENCLAVES";
    Errors["NFT_RECONCILIATION_FAILED"] = "NFT_RECONCILIATION_FAILED";
    Errors["RECONCILIATION_PAYLOAD_UNDEFINED"] = "RECONCILIATION_PAYLOAD_UNDEFINED";
})(Errors || (exports.Errors = Errors = {}));
