// ternoa-js/rent/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CancellationFeeAction = exports.RentFeeAction = exports.AcceptanceAction = exports.SubscriptionActionDetails = exports.DurationAction = void 0;
var DurationAction;
(function (DurationAction) {
    DurationAction["Fixed"] = "fixed";
    DurationAction["Subscription"] = "subscription";
})(DurationAction || (exports.DurationAction = DurationAction = {}));
var SubscriptionActionDetails;
(function (SubscriptionActionDetails) {
    SubscriptionActionDetails["PeriodLength"] = "periodLength";
    SubscriptionActionDetails["MaxDuration"] = "maxDuration";
    SubscriptionActionDetails["IsChangeable"] = "isChangeable";
    SubscriptionActionDetails["NewTerms"] = "newTerms";
})(SubscriptionActionDetails || (exports.SubscriptionActionDetails = SubscriptionActionDetails = {}));
var AcceptanceAction;
(function (AcceptanceAction) {
    AcceptanceAction["AutoAcceptance"] = "autoAcceptance";
    AcceptanceAction["ManualAcceptance"] = "manualAcceptance";
})(AcceptanceAction || (exports.AcceptanceAction = AcceptanceAction = {}));
var RentFeeAction;
(function (RentFeeAction) {
    RentFeeAction["Tokens"] = "tokens";
    RentFeeAction["NFT"] = "nft";
})(RentFeeAction || (exports.RentFeeAction = RentFeeAction = {}));
var CancellationFeeAction;
(function (CancellationFeeAction) {
    CancellationFeeAction["FixedTokens"] = "fixedTokens";
    CancellationFeeAction["FlexibleTokens"] = "flexibleTokens";
    CancellationFeeAction["NFT"] = "nft";
    CancellationFeeAction["None"] = "None";
})(CancellationFeeAction || (exports.CancellationFeeAction = CancellationFeeAction = {}));
